### ISSUE TITLE
Stop send Authorization header in the middle of push

### ISSUE THREAD

**Issue by farmerworking** [docker/docker#17104]
### System environment
docker info

	Containers: 6
	Images: 212
	Storage Driver: aufs
	 Root Dir: /var/lib/docker/aufs
	 Backing Filesystem: extfs
	 Dirs: 225
	 Dirperm1 Supported: true
	Execution Driver: native-0.2
	Logging Driver: json-file
	Kernel Version: 3.16.0-41-generic
	Operating System: Ubuntu 14.04.2 LTS
	CPUs: 1
	Total Memory: 993.8 MiB
	Name: i-m903up73
	ID: ZMBV:RGFO:PTY2:Z45B:5N2B:GDFX:NNL2:26ID:XJ4Y:U3K3:642J:2OQV
	Debug mode (server): true
	File Descriptors: 27
	Goroutines: 42
	System Time: 2015-10-16T15:00:18.902790463+08:00
	EventsListeners: 0
	Init SHA1: 70b0cda2ecbf0d9fb1879ed198a66ad09a3dc50c
	Init Path: /usr/lib/docker/dockerinit
	Docker Root Dir: /var/lib/docker
	WARNING: No swap limit support
	
docker version

	Client:
	 Version:      1.8.3
	 API version:  1.20
	 Go version:   go1.4.2
	 Git commit:   f4bf5c7
	 Built:        Mon Oct 12 05:37:18 UTC 2015
	 OS/Arch:      linux/amd64
	
	Server:
	 Version:      1.8.3
	 API version:  1.20
	 Go version:   go1.4.2
	 Git commit:   f4bf5c7
	 Built:        Mon Oct 12 05:37:18 UTC 2015
	 OS/Arch:      linux/amd64
	 
uname -a

	Linux i-m903up73 3.16.0-41-generic #55~14.04.1-Ubuntu SMP Sun Jun 14 18:43:36 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
	
docker-compose version

	docker-compose version: 1.4.2
	docker-py version: 1.3.1
	CPython version: 2.7.6
	OpenSSL version: OpenSSL 1.0.1f 6 Jan 2014

### steps to reproduce the bug
1. create a self-signed certificates, set the domain to app.example.com
2. insert 127.0.0.1 app.example.com to /etc/hosts
3. create a directory structure like that
![image](http://work.image.alimmdn.com/%E5%B1%8F%E5%B9%95%E5%BF%AB%E7%85%A7%202015-10-16%20%E4%B8%8B%E5%8D%883.09.26.png?t=1444979413000)

4. docker-compose.yml's content:
![image](http://work.image.alimmdn.com/%E5%B1%8F%E5%B9%95%E5%BF%AB%E7%85%A7%202015-10-16%20%E4%B8%8B%E5%8D%883.13.49.png?t=1444979644000)

5. config nginx as proxy to registry and do ssl termination for registry
6. auth_config.yml's content:
![image](http://work.image.alimmdn.com/%E5%B1%8F%E5%B9%95%E5%BF%AB%E7%85%A7%202015-10-16%20%E4%B8%8B%E5%8D%883.16.08.png?t=1444979784000)

I explain this architecture a little bit. Nginx will serve as a proxy to registry and do ssl termination for registry. registry is configured to use token based authentication where dockerauth plays the auth server role. 

The auth_config.yml is used to config dockerauth service. All that's matter in this case is that user with username: admin, password: 123, has all rights to do anything

#### the result I expected is:

1. login success
2. push success

#### However, the result actually is:

1. login success
2. push failed

![image](http://work.image.alimmdn.com/%E5%B1%8F%E5%B9%95%E5%BF%AB%E7%85%A7%202015-10-16%20%E4%B8%8B%E5%8D%883.20.56.png?t=1444980076000)

#### Additional information 1
I configure the nginx to log format as following which record the http_authorization header for me:

	log_format  main  '== $http_authorization == $remote_addr - $remote_user [$time_local] "$request" '
	
And the log about the push action shows
![image](http://work.image.alimmdn.com/%E5%B1%8F%E5%B9%95%E5%BF%AB%E7%85%A7%202015-10-16%20%E4%B8%8B%E5%8D%883.24.39.png?t=1444980332000)

You can see the first one has no authorization header which is expected. the following two request contains the token returned by dockerauth and be permitted.

The wired thing happend when the fourth request comes. All of sudden, the client decides to not send authorization header which cause the push action failed

#### Additional information 2
What's more wired is that once you stop use nginx as proxy and asking registry handle https himself which results following docker-compose.yml file

![image](http://work.image.alimmdn.com/%E5%B1%8F%E5%B9%95%E5%BF%AB%E7%85%A7%202015-10-16%20%E4%B8%8B%E5%8D%883.31.32.png?t=1444980705000)

with this settings, you can login and push successfully

**Comment by cpuguy83** [comment]
If you can login and push successfully when you take nginx out of the mix, this tells me that nginx is misconfigured.

**Comment by farmerworking** [comment]
@cpuguy83 I have to say that's a reasonable guess. But how can you explain the following log from nginx. Why would docker client send `Authorization` header in the second and third request, and stop sending it for the fourth request? What kind of misconfiguration would do that?

![image](http://work.image.alimmdn.com/%E5%B1%8F%E5%B9%95%E5%BF%AB%E7%85%A7%202015-10-16%20%E4%B8%8B%E5%8D%883.24.39.png?t=1444980332000)

I paste my nginx.conf file and hope it will help. One thing need to mention is that I use ngx_openresty bundle which enable nginx with lua ability

	worker_processes  auto;
	
	env REGISTRY_PORT_5000_TCP_ADDR;
	env REGISTRY_PORT_5000_TCP_PORT;
	
	error_log logs/error.log;
	pid        /var/run/nginx.pid;
	
	events {
	    worker_connections  1024;
	}
	
	
	http {
	    #include       /etc/nginx/mime.types;
	    default_type  application/octet-stream;
	
	    log_format  main  '$remote_addr - $remote_user [$time_local] "$request" '
	                      '$status $body_bytes_sent "$http_referer" '
	                      '"$http_user_agent" "$http_x_forwarded_for"';
	
	    access_log  logs/access.log main;
	
	    server_names_hash_bucket_size 64;
	
	    sendfile        on;
	    #tcp_nopush     on;
	
	    keepalive_timeout  65;
	
	    #gzip  on;
	
	    #include /etc/nginx/conf.d/*.conf;
	
	    server {
	        set_by_lua $addr 'return os.getenv("REGISTRY_PORT_5000_TCP_ADDR")';
	        set_by_lua $port 'return os.getenv("REGISTRY_PORT_5000_TCP_PORT")';
	
	        listen 80 default_server;
	
	        listen 443 ssl;
	
	        root /usr/share/nginx/html;
	        index index.html index.htm;
	
	        server_name your_domain.com;
	        ssl_certificate /ssl/domain.crt;
	        ssl_certificate_key /ssl/domain.key;
	
	        location / {
	            # try_files $uri $uri/ =404;
		        proxy_pass                       http://$addr:$port;
		        proxy_set_header  Host           $http_host;   # required for docker client's sake
		        proxy_set_header  X-Real-IP      $remote_addr; # pass on real client's IP
		        proxy_read_timeout               900;
	        }
	    }
	}

**Comment by endophage** [comment]
@farmerworking I'm not familiar with where in a request-response cycle nginx does its logging. Is it possible that `ngx_openresty` doesn't handle the PATCH method correctly and is stripping the `http_authorization` header, then logging is happening on that altered data?

Are you able to remove just the `ngx_openresty` plugin and test again or does it perform a critical operation (even if it is critical to your production deploy this would still be a useful test to perform to determine if this is indeed the problem)?

**Comment by farmerworking** [comment]
![image](http://work.image.alimmdn.com/%E5%B1%8F%E5%B9%95%E5%BF%AB%E7%85%A7%202015-10-27%20%E4%B8%8B%E5%8D%883.29.08.png?t=1445931151307)

I walk into another issue which I think it's relavent to this. 

In the log I pasted. You can see the first 2 requests is using `https` which can be handled correctly. In the third request, the `patch` request, it's using `http` instead which results an `400 bad request` error

This happened when you use nginx to reverse proxy request to private registry and it's gone when you configure the private registry to handle https himself

It's gone doesn't necessary indicating the `patch` request is using https again. It just means you can successfully push a image

I strongly suggest looking into where `patch` method is invoked. All wired things happened with this request in the middle. It seems like this request is not following the same rules the others do

**Comment by farmerworking** [comment]
![image](http://work.image.alimmdn.com/%E5%B1%8F%E5%B9%95%E5%BF%AB%E7%85%A7%202015-10-27%20%E4%B8%8B%E5%8D%883.29.08.png?t=1445931151307)

I walk into this issue today but I think it's relevant to this issue. This time the first two requests using `https` method and worked correctly, in the third request it changed to using `http` instead and failed because of 400 bad request error. This happened when I use nginx to handle https and reverse proxy to private registry. But once I remove nginx and ask private registry to handle https itself the problem is gone.

What I mean it's gone doesn't indicate that the `patch` request will use `https` in the latter case. It just means the image can be pushed into private registry successfully

Since all the wired things happen in the middle and in this particular http method `patch`, I strongly suggest looking into where this request happened. It seems behave differently from the others

@endophage this time I do not use the `ngx_openresty` bundle but install nginx from official source and manually configure it with lua module. You can suspect nginx may drop several header fields. It's unlikely nginx will change your https request to http request and report 400 bad request error about it. Given the similarity of these two scenarios, I think the problem lies in docker itself

**Comment by endophage** [comment]
@farmerworking as the logging includes response codes, nginx is clearly logging after any modules have been applied. If any modules do not handle the `PATCH` method correctly, given you have nginx doing ssl termination (which has to happen before the data is passed to any other modules), I still strongly suspect that it is the nginx configuration or one of the modules that is causing the problem. 

It's interesting that when you switch from `ngx_openresty` to the other lua module, it changed from a 401 to a 400 response. That makes me more suspicious that the problem lies in nginx or your lua code. Could you pleease run a test with just plain nginx, no lua modules at all.

I took a look at the code sending the `PATCH` requests and it appears to be building requests from the same base URL as the other requests. Particularly the URL used for the `POST` immediately preceding the `PATCH` is the URL actually used to initialize the httpBlobUpload object which is responsible for sending the `PATCH` requests: https://github.com/docker/docker/blob/master/vendor/src/github.com/docker/distribution/registry/client/repository.go#L436

**Comment by farmerworking** [comment]
@endophage What you said sounds reasonable enough so I try the official nginx image instead. The log is that:

![image](http://work.image.alimmdn.com/%E5%B1%8F%E5%B9%95%E5%BF%AB%E7%85%A7%202015-10-28%20%E4%B8%8A%E5%8D%889.36.45.png?t=1445996295770)

You can see in the third request, no `http authorization` header is carried and `http` is used

**Comment by farmerworking** [comment]
I create a really simple demo which can reproduce the bug

### file structure

![image](http://work.image.alimmdn.com/%E5%B1%8F%E5%B9%95%E5%BF%AB%E7%85%A7%202015-10-28%20%E4%B8%8A%E5%8D%889.47.19.png?t=1445996864325)

### docker-compose.yml content:

	proxy:
	  image: nginx
	  volumes:
	    - .:/ssl
	    - .:/etc/nginx
	  ports:
	    - "5001:443"
	registry:
	  image: registry:2
	  ports:
	    - "5000:5000"

### nginx.conf content:

	user  nginx;
	worker_processes  1;
	
	error_log  /var/log/nginx/error.log warn;
	pid        /var/run/nginx.pid;
	
	
	events {
	    worker_connections  1024;
	}
	
	
	http {
	    default_type  application/octet-stream;
	
	    log_format  main  '== $http_authorization == $scheme == $remote_addr - $remote_user [$time_local] "$request" '
	                      '$status $body_bytes_sent "$http_referer" '
	                      '"$http_user_agent" "$http_x_forwarded_for"';
	
	    access_log  /var/log/nginx/access.log  main;
	
	    sendfile        on;
	    #tcp_nopush     on;
	
	    keepalive_timeout  65;
	
	    #gzip  on;
	
	    include /etc/nginx/conf.d/*.conf;
	
	    server {
	        listen       443 ssl;
	
	        ssl_certificate      /ssl/domain.crt;
	        ssl_certificate_key  /ssl/domain.key;
	
	        ssl_session_cache    shared:SSL:1m;
	        ssl_session_timeout  5m;
	
	        ssl_ciphers  HIGH:!aNULL:!MD5;
	        ssl_prefer_server_ciphers  on;
	
	        location / {
	          proxy_pass                       http://121.201.58.179:5000;
	          proxy_set_header  Host           $http_host;   # required for docker client's sake
	          proxy_set_header  X-Real-IP      $remote_addr; # pass on real client's IP
	          proxy_read_timeout               900;
	        }
	    }
	}

### steps
1. create a self-assigned certificate file
2. docker tag hello-world example.com:5001/hello-world && docker push example.com:5001/hello-world

Use 5001 instead of default 443 for https so that you can see more clearly from nginx log file

If you use 443 port, the docker client will report `dial tcp 121.201.58.179:80: connection refused` because it try to send `patch` request using `http`

**Comment by farmerworking** [comment]
one thing to notice:

    docker daemon ===> https ===> nginx ===> http ===> registry ------- not work

    docker daemon ===> https ===> nginx ===> https ===> registry ------- work

Certainly registry has something to do with that

**Comment by endophage** [comment]
@farmerworking apologies, I've been wrapped up in prep for DockerCon EU. I'll probably get a chance to dig into your test and look at things in more detail in a week or two.

**Comment by half-dead** [comment]
@endophage @farmerworking Any updates? We've run into the same situation.

**Comment by half-dead** [comment]
@farmerworking So glad to tell you that we've solved this problem, just add this to your nginx.conf
```
proxy_set_header  X-Forwarded-Proto $scheme;
```

PS：If there is multiple layers of nginx in your system, you only need to add this header in the first layer; otherwise, it might be overwritten by following layers.

**Comment by thaJeztah** [comment]
@half-dead @endophage does this need documentation changes?

**Comment by endophage** [comment]
@half-dead thanks for finding a solution! Also (for entirely selfish reasons) great to hear it's on the nginx side.

@thaJeztah yes, this should probably be added to (this documentation)[https://github.com/docker/distribution/blob/master/docs/nginx.md] and there may be other places nginx setup is discussed.

**Comment by thaJeztah** [comment]
@half-dead since you have a working config, are you interested in creating a pull request for updating the documentation?

**Comment by half-dead** [comment]
@endophage  It's already covered in [here](https://github.com/docker/distribution/blob/master/docs/nginx.md#gotchas), but still, I did a minor modification.

**Comment by thaJeztah** [comment]
@half-dead ah, I actually remembered that being added, LOL

With that information, and the documentation in place, I think we can close this issue, but feel free to comment here if you think I closed this by mistake :+1:
